pkg/receive/handler.go

```diff
@@ -202,7 +203,7 @@ func NewHandler(logger log.Logger, o *Options) *Handler {
 	ins := extpromhttp.NewNopInstrumentationMiddleware()
 	if o.Registry != nil {
 		ins = extpromhttp.NewTenantInstrumentationMiddleware(
-			o.TenantHeader,
+			path.Base(o.TenantHeader),
```

Comment:
So there are a couple of things here. I don't think this is the right place/input to sanitize, as this is the HTTP header name (by default `THANOS-TENANT`) and this is passed to the Prometheus registry for collecting metrics. We would need to sanitize the value of this HTTP header instead.
 
To highlight the full flow, this header value is [fetched](https://github.com/thanos-io/thanos/blob/main/pkg/receive/handler.go#LL393) in the `receiveHTTP` handler, and then goes through `handleRequest` -> `forward` -> `fanoutForward` and passed to [`h.writer.Write()`](https://github.com/thanos-io/thanos/blob/main/pkg/receive/handler.go#L655) which creates a multiTSDB [`TenantAppendable`](https://github.com/thanos-io/thanos/blob/main/pkg/receive/writer.go#L60) to write data for that tenant. There is some logic ([`getOrLoadTenant`](https://github.com/thanos-io/thanos/blob/main/pkg/receive/multitsdb.go#L542)) around how a tenant TSDB is started, but the HTTP header value is eventually used to decide dataDir for the tenant using the [`defaultTenantDataDir`](https://github.com/thanos-io/thanos/blob/main/pkg/receive/multitsdb.go#L538) method.

So, we can sanitize this in `receiveHTTP` itself, but I wonder if we should treat tenant id like a path at all. If tenant header value is set to something like `///`, `path.Base` would return `/`. Instead maybe we should use some regex that fails if tenant id has some backward/forward slash. WDYT? 🙂 

Also, I think we'd need to sanitize [`--receive.default-tenant-id`](https://github.com/thanos-io/thanos/blob/main/cmd/thanos/receive.go#L849) in a similar way.

Reply:
What do you think about having an aliased type that would be used for this? For example:

```go
type tenantID string

func (t *tenantID) String() string {
  return string(*t)
}

func NewTenantID(s string) *tenantID {
  s = path.Base(s)
  return &tenantID{s}
}
```

(untested but shows my idea). This way, we would have the cleaning-up logic only in one place.

Reply:
Yup, I think this would make it way cleaner!

Reply:
Hi all I think that using `path.Base` is only safe at configuration parse-time. Any time this is used at run-time we are liable to run into very sketchy situations where we think we are safe but we actually aren't. Consider for example nice tenant `A` and malicious tenant `B/A`. If we `path.Base` at run-time and let tenant `B/A` write data to disk, they will be writing into tenant `A`'s directory.

This means that we cannot just rely on doing `path.Base(tenantHeaderValue)` to find the directory to write to. We actually have to test `path.Base(tenantHeaderValue) == tenantHeaderValue` and return a 503 or something if not.

Reply:
Hi o/ thanks for the review! totally, I didn't think of that, but deff can be used for overriding a tenant :\